server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1234,13 +1235,19 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
         context.size(source.size());
         Map<String, InnerHitContextBuilder> innerHitBuilders = new HashMap<>();
         QueryBuilder query = source.query();
+        InnerHitsRewriteContext innerHitsRewriteContext = new InnerHitsRewriteContext(
+            context.getSearchExecutionContext().getParserConfig(),
+            context::getRelativeTimeInMillis
+        );
         if (query != null) {
-            InnerHitContextBuilder.extractInnerHits(query, innerHitBuilders);
+            QueryBuilder rewrittenForInnerHits = Rewriteable.rewrite(query, innerHitsRewriteContext, true);
```

Comment:
I think we should do instead:
`QueryBuilder query = Rewriteable.rewrite(query, innerHitsRewriteContext, true);`

Otherwise we again need to rewrite the query on line 1246.

Reply:
I don't think so. These are separate rewrites. One is only for inner-hits extraction, the other is for querying. These could match different documents. The queries that are rewritten to between them could be different.

Reply:
Where does rewrite for inner hits happen?  It seems to me that it happens both in QueryPhase and FetchPhase. Does it mean that we run exact knn query two times, basically computing scores for all inner hits two times?

Reply:
> Where does rewrite for inner hits happen? It seems to me that it happens both in QueryPhase and FetchPhase. Does it mean that we run exact knn query two times, basically computing scores for all inner hits two times?

Inner hit query execution only happens during fetch. However no matter the phase, `InnerHitContextBuilder.extractInnerHits(query, innerHitBuilders);` is called. We should have a larger rewrite where `InnerHitContextBuilder` only occurs during the fetch phase.

This is why I don't use the `rewrittenForInnerHits` query anywhere but within `extractInnerHits`.